provider/kubernetes_test.go

```diff
@@ -1524,11 +1524,80 @@ func TestServiceAnnotations(t *testing.T) {
 	}
 }
 
+func TestGetServiceErrors(t *testing.T) {
+	ingresses := []*v1beta1.Ingress{
+		{
+			ObjectMeta: v1.ObjectMeta{
+				Namespace: "testing",
+				Annotations: map[string]string{
+					"traefik.frontend.passHostHeader": "true",
+				},
+			},
+			Spec: v1beta1.IngressSpec{
+				Rules: []v1beta1.IngressRule{
+					{
+						Host: "foo",
+						IngressRuleValue: v1beta1.IngressRuleValue{
+							HTTP: &v1beta1.HTTPIngressRuleValue{
+								Paths: []v1beta1.HTTPIngressPath{
+									{
+										Path: "/bar",
+										Backend: v1beta1.IngressBackend{
+											ServiceName: "notExists",
+											ServicePort: intstr.FromInt(80),
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	services := []*v1.Service{}
+	endpoints := []*v1.Endpoints{}
+	watchChan := make(chan interface{})
+
+	client := clientMock{
+		ingresses:       ingresses,
+		services:        services,
+		endpoints:       endpoints,
+		watchChan:       watchChan,
+		apiServiceError: true,
+	}
+	provider := Kubernetes{}
+	_, err := provider.loadIngresses(client)
+
+	if err == nil {
+		t.Fatal("Should have raised error when loading services but did not.")
```

Comment:
Instead of just checking for the error being `nil`, what I think we can do here is pass a custom error variable to the mock (e.g., `errors.New("failed")`) and check against this very error. It shouldn't matter what error value we inject since our code doesn't depend on any specific error -- we just want to verify that we return the error to the caller that the (mocked) API gives us.
Similar for the endpoint case.

I hope this addresses your third question.